### PR TITLE
Mixture of Experts

### PR DESCRIPTION
**Pull request by kernelmachine**
* Implements top-2 routing with arbitrary numbers of experts.

### CODE REVIEW

**Comment by sagadre** [comment]
i think those are maybe from an earlier PR. @kernelmachine u may need to merge with latest main

**Comment by kernelmachine** [comment]
Ooh okay will fix those merge conflicts! 

Regardless, I am also dealing with some mysterious bug here, as I am seeing worse perplexities with more experts, at a budget I expect the MoE to do better than the dense model on: https://docs.google.com/spreadsheets/d/1QrjOA24wDGGXyZgn2TsAI4q4nmdz5ceQXTTF7pt8hT4/edit#gid=0

File: open_lm/main.py [hunk 1]
```diff
@@ -399,7 +399,11 @@ def main(args):
     if args.hf_model is not None:
         model = create_wrapped_hf_model(args)
     else:
-        with torch.device("meta" if args.fsdp else args.device):
+        # with torch.device("meta" if args.fsdp else args.device):
+        #     model = create_model(args)
+        # if not args.fsdp:
+        #     model.reset_parameters()
+        with torch.device(args.device):
```
Comment (achalddave) [inline open_lm/main.py:406]: Note for later: before merging this pr, let's merge https://github.com/mlfoundations/open_lm/pull/155 and require --disable-meta-device for MoE

**Comment by kernelmachine** [comment]
This branch is ready to merge. Final benchmarking numbers:

 Here are the markdown tables matching the spreadsheet:

### Compute budgets

| Compute type | 41M  | 87M  | 160M | 410M | 830M |
|--------------|------|------|------|------|------|
| Number of nodes | 1   | 1    | 1    | 2 | 4 |
| Number of tokens | 20.0B | 20.0B | 20.0B | 20.0B | 20.0B | 

| Performance | Model ppl | toks/sec/gpu |
|-|-|-|
| Number of Experts | 41M | 87M | 160M | 410M | 830M | 410M, 1 node | 
| 1 | 27.61 | 18.68 | 14.87 | 10.54 | 9.39 | 830M, 1 node |  
| 8 | 19.85 | 14.66 | 12.26 | 9.82 | 8.84 |

### Tokens/sec/GPU 

| Number of Experts | 41M | 87M | 160M | 410M | 830M |
|:-:|:-:|:-:|:-:|:-:|:-:|
| 1 | 141.2K | 106.0K | 95.5K | 30.3K | 16.0K |  
| 8 | 69.5K | 66.6K | 66.2K | 18.5K | 9.2K |

### Training Parameters

| Number of Experts | 41M | 87M | 160M | 410M | 830M |
|-|-|-|-|-|-|
| 8 experts | 68.9M | 165.4M | 360.6M | 1.1B | 2.4B |
| 32 experts | 164.5M | 439.9M | 1.0B | 3.5B | 7.9B |

### Inference Parameters  

| Number of Experts | 41M | 87M | 160M | 410M | 830M |  
|-|-|-|-|-|-|
| 2 experts | 45.0M | 96.8M | 190.7M | 509.2M | 1.1B |

File: open_lm/main.py (hunk 1 above)
Comment (kernelmachine) [reply open_lm/main.py]: this is done!

File: open_lm/model.py [hunk 2]
```diff
@@ -19,6 +20,15 @@
 from open_lm.positional_embedding.rotary import RotaryWithCast
 from open_lm.positional_embedding.llama_rotary import LLaMARotaryWithCast
 
+
+# from open_lm.moe.mixture_of_experts import MoE
```
Comment (mitchellnw) [inline open_lm/model.py:24]: Optionally remove

File: open_lm/model.py [hunk 3]
```diff
@@ -19,6 +20,15 @@
...
+try:
+    from megablocks.layers.moe import MoE
+    from megablocks.layers.arguments import Arguments as MoEArgs
+except ImportError:
+    logging.warning(f"Megablocks not installed. To train MoE, install with pip install megablocks.")
```
Comment (mitchellnw) [inline open_lm/model.py:29]: Should we have some assert here to make sure they're not using MoE?

File: open_lm/train.py [hunk 4]
```diff
@@ -11,6 +11,13 @@
 from torch.distributed.distributed_c10d import ReduceOp
 from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
 
+
+try:
+    from megablocks.layers.moe import batched_load_balancing_loss, clear_load_balancing_loss
+    from megablocks.layers.arguments import Arguments as MoEArgs
+except ImportError:
+    logging.warning(f"Megablocks not installed. To train MoE, install with pip install megablocks.")
```
Comment (mitchellnw) [inline open_lm/train.py:19]: Again maybe some assert would be good?

File: open_lm/train.py [hunk 5]
```diff
@@ -228,11 +236,37 @@ def train_one_epoch(model, data, loss, epoch, step, optimizer, scaler, scheduler
...
+                        local_load_balancing_loss = batched_load_balancing_loss(moe_args)
+                        clear_load_balancing_loss()
+
+                        local_loss = local_load_balancing_loss + local_loss
+
                     backward(local_loss, scaler)
                 if ii == 0:
-                    total_loss = local_loss
+                    if args.moe_freq > 0:
+                        total_loss = local_loss - local_load_balancing_loss
```
Comment (mitchellnw) [inline open_lm/train.py:260]: Why do we subtract load balancing loss?

**Comment by mitchellnw** [comment]
Looks great! Left a bunch of minor comments but other than that ready to merge.

**Review by mitchellnw — APPROVED** [review]
(See above)

File: open_lm/train.py [hunk 6]
```diff
@@ -228,11 +236,37 @@ def train_one_epoch(model, data, loss, epoch, step, optimizer, scaler, scheduler
                             * inputs_ii.shape[0]
                             / inputs.shape[0]
                         )
+                    if args.moe_freq > 0:
+                        moe_args = MoEArgs(
```
Comment (achalddave) [inline open_lm/train.py:240]: Can we construct moeargs outside of the train loop to make the training loop smaller and avoid constructing this every batch?

File: open_lm/train.py (hunk 5 above)
Comment (achalddave) [reply open_lm/train.py:260]: Same question. I guess maybe this is so we can log the loss (without load balancing) separately from the load balancing loss? But if so, let's just subtract this outside of the grad accum loop when logging, instead?

File: open_lm/train.py [hunk 7]
```diff
@@ -228,11 +236,37 @@ def train_one_epoch(model, data, loss, epoch, step, optimizer, scaler, scheduler
...
+                            hidden_size=model.dim,
+                            ffn_hidden_size=model.dim * 4,
+                            moe_num_experts=args.moe_num_experts,
+                            num_layers=model.n_layers // 2,
+                            moe_expert_model_parallelism=True,
```
Comment (achalddave) [inline open_lm/train.py:245]: I think this only implements the MoE losses for the case with accumulation. Can you also implement it above for the case without accumulation? (Ideally we should merge these two in a cleaner way...)

**Review by achalddave — CHANGES_REQUESTED** [review]
This looks great! I left a few comments, particularly I think some MoE things need to be implemented in the case where accum_freq == 1.

File: open_lm/train.py (hunk 5 above)
Comment (kernelmachine) [reply open_lm/train.py]: yep that is correct @achalddave . I modified this to an "lm_loss" and "load balancing loss" and removed the subtraction so its clearer

File: open_lm/model.py (hunk 3 above)
Comment (kernelmachine) [reply open_lm/model.py]: we can't check args during imports though, right?

File: open_lm/train.py (hunk 6 above)
Comment (kernelmachine) [reply open_lm/train.py]: done

**Review by Vaishaal — APPROVED** [review]
this is my favorite PR, LGTM

File: open_lm/train.py [hunk 8]
```diff
@@ -268,13 +312,20 @@ def train_one_epoch(model, data, loss, epoch, step, optimizer, scaler, scheduler
...
-            losses_m.update(global_loss_tensor.item(), batch_size)
+            if args.moe_freq > 0:
+                losses_m.update(global_loss_tensor.item() - total_load_balancing_loss.item(), batch_size)
+                load_balancing_losses_m.update(total_load_balancing_loss.item(), batch_size)
+            else:
+                losses_m.update(global_loss_tensor.item(), batch_size)
             samples_per_second = inputs.numel() * args.world_size / batch_time_m.val
             samples_per_second_per_gpu = inputs.numel() / batch_time_m.val
+            loss_str = f"Loss: {losses_m.avg:.3f}"
+            loss_str += f" LB-Loss: {load_balancing_losses_m.avg:.3f}" if args.moe_freq > 0 else ""
```
Comment (sagadre) [inline open_lm/train.py:325]: maybe guard this with a args.moe_freq > 0 ? also might be good do do a open_lm_160m run with this branch to make sure nothing in non-moe codepath affected

**Comment by sagadre** [comment]
lgtm! might be good do do a open_lm_160m run with this branch to make sure nothing in non-moe codepath affected

**Review by achalddave — APPROVED** [review]

**Comment by kernelmachine** [comment]
Confirmed same training losses with main branch
